### ISSUE TITLE
Problem with property_enum_convention rule — PROP_0 is an outdated convention

### ISSUE THREAD

**Issue by pwithnall** [bilelmoussaoui/gobject-lint#52]
Going by what `RULES.md` says (I haven’t had a chance to run goblint yet, sorry, but it does look exciting), the `property_enum_convention` rule is outdated.

The current best practice that I’ve been using is:
```
typedef enum
{
  PROP_FOO = 1,
  PROP_BAR,
  PROP_BAZ,
} MyObjectProperty;
```

Note there’s no `PROP_0` and there’s no `N_PROPS` element either.

This allows the `prop_id` to be cast to `(MyObjectProperty) prop_id` in `get_property()` and `set_property()` which allows `-Wswitch-enum` to be used to warn about properties accidentally not handled in those vfuncs. It also means that `g_object_class_install_properties()` can be called with `G_N_ELEMENTS (props)` to install the correct number of properties without needing to use `N_PROPS`.

See, for example, https://gitlab.freedesktop.org/pwithnall/malcontent/-/blob/main/libmalcontent-web/filter-updater.c?ref_type=heads#L114 for some recently written code which uses this approach.

**Comment by bilelmoussaoui** [comment]
The rule now automatically ports to this more modern pattern.

**Comment by pwithnall** [comment]
\o/

Skimming through the changes, `// Remove trailing comma from previous property` looks a bit odd to me. I’d keep a trailing comma on each line to reduce diff noise when new properties are added.

And one of the key advantages of this approach is being able to enable `-Wswitch-enum`, but that’s only possible if the `switch` statements in `get_property()` and `set_property()` cast their `int prop_id` argument to the `MyObjectProperty` enum type. Then the compiler can detect missing `case` statements.

Tell me to shut up if I’ve misinterpreted the changes though! 😅

**Comment by bilelmoussaoui** [comment]
The rule now also changes get_property/set_property and make use of the enum but we have a case like
```c
static void
meta_screen_cast_session_set_property (GObject      *object,
                                       guint         prop_id,
                                       const GValue *value,
                                       GParamSpec   *pspec)
{
  MetaScreenCastSession *session = META_SCREEN_CAST_SESSION (object);

  switch (prop_id)
    {
    case PROP_REMOTE_DESKTOP_SESSION:
      session->remote_desktop_session = g_value_get_object (value);
      if (session->remote_desktop_session)
        session->session_type = META_SCREEN_CAST_SESSION_TYPE_REMOTE_DESKTOP;
      else
        session->session_type = META_SCREEN_CAST_SESSION_TYPE_NORMAL;
      break;

    case N_PROPS + META_DBUS_SESSION_PROP_SESSION_MANAGER:
      session->session_manager = g_value_get_object (value);
      break;
    case N_PROPS + META_DBUS_SESSION_PROP_PEER_NAME:
      session->peer_name = g_value_dup_string (value);
      break;
    case N_PROPS + META_DBUS_SESSION_PROP_ID:
      session->session_id = g_value_dup_string (value);
      break;
    default:
      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
      break;
    }
}
```
for implemented ifaces properties, how would that needs to be handled?

**Comment by bilelmoussaoui** [comment]
For now we skip the case I mentioned but otherwise we port the get_property/set_property to use a casted prop_id, works great on a complex code base like mutter.

**Comment by blandford** [comment]
So, I ran this with `--fix` and ended up with a crasher due to a PROP_ORIENTATION overrride.

The old code had 
```C
enum {
  PROP_0,
  PROP_VALID,
  PROP_SPACING,
  PROP_PUZZLE_KIND,
  PROP_ORIENTATION,
  N_PROPS = PROP_ORIENTATION
};

static GParamSpec *obj_props[N_PROPS] = { NULL, };
```

Fixed  code looks like
```C
{
   PROP_VALID = 1,
   PROP_SPACING,
   PROP_PUZZLE_KIND,
   PROP_ORIENTATION
};

static GParamSpec *obj_props[PROP_ORIENTATION + 1] = { NULL, };
``` 
Which leaves a NULL in the ORIENTATION spot. Ideally, goblint could capture  `g_object_class_override_property()` and not include those in the N_PROPS, or capture when the old array wasn't actually the full set of properties.

**Comment by bilelmoussaoui** [comment]
I am not sure yet how to handle properties override, there is no clear way to detect them as the pattern i noticed was 

```
enum {
  PROP_0,
  PROP_VALID,
  PROP_SPACING,
  PROP_PUZZLE_KIND,
  PROP_ORIENTATION,
  N_PROPS,

  PROP_ORIENTATION
};
```

instead of what you currently have...

**Comment by blandford** [comment]
Hmmm. I took my approach from `GtkBox`.

Ultimately, the old code has an enum value for obj_props[]; maybe you can key off of that?

**Comment by bilelmoussaoui** [comment]
but then we can't cast prop_id to MyObjectProps no?

**Comment by bilelmoussaoui** [comment]
I have added overridden properties support, the only missing case is implemented ifaces properties, but we skip those for now so there should be no mis-generated fixes.

**Comment by swick** [comment]
This is a bit weird because it's not an obvious improvement. It trades one kind of safety for another one:

* With `-Wswitch-enum` the compiler will yell at you if you forget to handle a variant, but you have to remember to update the last property in `obj_props`
*  With `N_PROPS` you don't have to remember to change `obj_props` but you don't get `-Wswitch-enum`

**Comment by bilelmoussaoui** [comment]
I guess we can keep N_PROPS and handle it in get_property/set_property with
```


static void
mct_filter_updater_set_property (GObject      *object,
                                 guint         property_id,
                                 const GValue *value,
                                 GParamSpec   *pspec)
{
  MctFilterUpdater *self = MCT_FILTER_UPDATER (object);

  switch ((MctFilterUpdaterProperty) property_id)
    {
    case PROP_POLICY_MANAGER:
      /* Construct only. */
      g_assert (self->policy_manager == NULL);
      self->policy_manager = g_value_dup_object (value);
      break;
    case PROP_USER_MANAGER:
      /* Construct only */
      g_assert (self->user_manager == NULL);
      self->user_manager = g_value_dup_object (value);
      break;
    case PROP_STATE_DIRECTORY:
      /* Construct only. */
      g_assert (self->state_directory == NULL);
      self->state_directory = g_value_dup_object (value);
      break;
    case PROP_CACHE_DIRECTORY:
      /* Construct only. */
      g_assert (self->cache_directory == NULL);
      self->cache_directory = g_value_dup_object (value);
      break;
    case N_PROPS:
      g_assert_not_reached ();
    }
}
```
would that be good enough?

**Comment by pwithnall** [comment]
I disagree with `N_PROPS` existing: semantically, it should not be a member of the enum, as it’s not a property of the class. It’s data about the properties enum which we happen to put in the enum for convenience.

The compiler or asan should warn if `PROP_FOO + 1` doesn’t get updated when a new property is added, as it fairly quickly results in an out-of-bounds write when the class is inited. I’ve never found incrementing this to be a problem in practice.

Sebastian is correct in that this approach isn’t perfect (ideally we need an operator in C which tells you the largest value of an enum type), but it does make more errors more detectable:

> * With `-Wswitch-enum` the compiler will yell at you if you forget to handle a variant, but you have to remember to update the last property in `obj_props`

If you forget to update the last property in `obj_props`, class init will crash, and some compilers will tell you at compile time that you’re assigning off the end of the `obj_props` array. So errors will be detected at class init time in the worst case.

> * With `N_PROPS` you don't have to remember to change `obj_props` but you don't get `-Wswitch-enum`

In this case, errors are only detected when you get/set the property which is missing from the `switch` statements, which happens more rarely than class init.

**Comment by bilelmoussaoui** [comment]
I guess we can make the linter detect that the last property is the one used in the array ?

**Comment by pwithnall** [comment]
wfm!

**Comment by swick** [comment]
It's not ideal, but I don't think there is an ideal solution here, so that works for me.

**Comment by bilelmoussaoui** [comment]
People in gnome-shell channel didn't like the new style suggested here so i kept both and you can opt-out of the typed style by using 

```toml
[rules.property_enum_convention]
style = "legacy"
```

For now, let us close this until there is something actionable we can do.
